Feature/string enum mspec

hutcheb opened the pull request:
To be used in conjunction with the build-tools feature/opcua_native_client branch.

- Implemented integration tests for the Java, Go and C build-util modules. 
    - This takes the test.mspec in the protocol-test module and generates code for each language
    - It then tries to compile these sources. (Ideally it would be good implement tests for the parsing/serializing but should probably wait for the test.mspec to be improved upon beforehand)

- Implemented Enumerated Discriminators in mspec, e.g. `[discriminator EnumType 'discr']`
    - This avoids us working around it by not assigning the discriminator as a simple field and having to parse ti as a parameter when creating objects.

- An accidental but slight improvement was added when defining Enumerated constants and switch fieds, you no longer have to include the Enumerated Type for each case e.g.
```
    [typeSwitch 'discr'
        ['EnumType.BOOL'  DiscriminatedBOOL]
```
becomes
```
    [typeSwitch 'discr'
        ['BOOL'  DiscriminatedBOOL]
```
- Implemented variable lengths for string fields e.g.
    `[simple string '8' 'UTF-8' 'stringField']`
    expressions can also be used
    `[simple int 8 'stringLength']`
    `[simple string 'stringLength' 'UTF-8' 'stringField']` // Have to update the templates when using implicit fields for stringLengths Ill do that tomorrow.

- Added exclusions for RAT checks in C so that maven can be used to compile it.

- Added the Go and C build-util modules to their respective profiles, with-go and with-c so we don't build them when only using java.

hutcheb committed:
Add support for variable strings and enum discriminators.

hutcheb committed:
Integration test works with existing test protocol.

Need to remove my test protocol I added
Add Integration test to other protocols
Fill out the test protocol with valid mspec that covers all documented types.

hutcheb committed:
Cleaned up build-utils integration tests.

Still have an issue with building the c and go modules
For C it relies on the spi module but hasn't been compiled yet.

hutcheb committed:
Compiled the test protocol mspec in C

Checksum support is missing
Updated the int and uint type selection so they match Java.
there was an issue when selecting non standard sized types.

hutcheb committed:
Got enumerator discriminators working for for C

hutcheb committed:
Updated C and Go templates

Still have to confirm Go generated sources ae being compiled to check for some errors.
Have to rework the java templates as the mspec format changed when implementing C

hutcheb committed:
Updated Enum Discriinator fields for C and Java.

hutcheb committed:
Merge branch 'develop' into feature/string_enum_mspec

hutcheb committed:
Integration test is working in Go, sorting through some failed tests.

hutcheb committed:
Minor changes

hutcheb committed:
Integration tests working for java,C and Go

hutcheb committed:
Add additional tests to the test.mspec

sruehl commented on .gitignore:
```diff
@@ -181,3 +181,14 @@ build-iPhoneSimulator/
+*.cmake
+*.internal
+link.txt
+Makefile
```
```suggestion
```
This would actually remove the plc4go "wrapper" Makefile from plc4gp again so a more specific ignore would be good. (Not sure why the Makefile is ignored here)

hutcheb replied:
The makefiles in plc4c are generated using cmake. I have changed it to only ignore plc4c makefiles.

sruehl commented on build-utils/language-c/pom.xml:
```diff
@@ -32,6 +32,40 @@
+          <debug>true</debug>
+          <projectsDirectory>src/test/resources</projectsDirectory>
+          <cloneProjectsTo>${project.build.directory}/integration-tests</cloneProjectsTo>
+          <!-- Removed so that a local version of the build-tools could be used. Need to add it back in -->
```
Guess this needs to be done before merging?

sruehl commented on build-utils/language-c/pom.xml:
```diff
@@ -55,4 +89,5 @@
     </dependency>
   </dependencies>
 
+
```
Accidental newline?
```suggestion
```

sruehl commented on build-utils/language-c/src/main/java/org/apache/plc4x/language/c/CLanguageTemplateHelper.java:
```diff
@@ -178,33 +178,44 @@ public boolean requiresPointerAccess(ComplexTypeDefinition typeDefinition, Strin
      */
     @Override
     public String getLanguageTypeNameForTypeReference(TypeReference typeReference) {
+        System.out.println("Type Reference " + typeReference.getClass());
```
```suggestion
```
Debug statement I guess (We might add more logging to the whole language helper in the future I reckon).

sruehl commented on build-utils/language-c/src/main/java/org/apache/plc4x/language/c/CLanguageTemplateHelper.java:
```diff
@@ -178,33 +178,44 @@ public boolean requiresPointerAccess(ComplexTypeDefinition typeDefinition, Strin
                 case BIT:
                     return "bool";
-                case UINT:
+
```
```suggestion
```
accidental newline?

sruehl commented on build-utils/language-c/src/main/java/org/apache/plc4x/language/c/CLanguageTemplateHelper.java:
```diff
@@ -288,7 +299,7 @@ public String getTypeSizeForField(TypedField field) {
     public String escapeValue(TypeReference typeReference, String valueString) {
         if (valueString == null) {
-            return null;
+            return "NULL";
```
Shouldn't that than be a empty string ""?
```suggestion
            return "";
```
(Dislaimer: not a C expert ;)

hutcheb replied:
At this point it could be any string, this only gets used once in enum-template-c line 95 and it just checks for a "-1" value.
As long as it doesn't return an actual null we should be good.

sruehl commented on build-utils/language-c/src/main/java/org/apache/plc4x/language/c/CLanguageTemplateHelper.java:
```diff
@@ -380,7 +390,7 @@ public String getReadBufferReadMethodCall(SimpleTypeReference simpleTypeReferenc
             case STRING:
                 StringTypeReference stringTypeReference = (StringTypeReference) simpleTypeReference;
-                return "plc4c_spi_read_string(io, " + stringTypeReference.getSizeInBits() + ", \"" +
+                return "plc4c_spi_read_string(io, " + stringTypeReference.getLength() + " * 8, \"" +
```
What is the reason using bytes sized lengths here? Isn't that supported for strings?

hutcheb replied:
I added a getLength (now getLengthExpression) to the StringTypeReference class as it now uses an expression instead of a hard coded value.
getLengthExpression returns the size in bits, I have removed the "* 8"

sruehl commented on build-utils/language-c/src/main/resources/templates/c/data-io-template-c.ftlh:
```diff
@@ -67,7 +67,7 @@
 #include <plc4c/utils/list.h>
 #include <plc4c/spi/evaluation_helper.h>
-#include <plc4c/driver_${helper.getProtocolName()}.h>
+
```
could be accidental newline but I guess this is the C-like import grouping? Would than expect one after time.h too

sruehl commented on build-utils/language-c/src/main/resources/templates/c/enum-template-h.ftlh:
```diff
@@ -50,6 +50,7 @@ ${helper.getIncludesDirectory()?replace(".", "/")}/${helper.camelCaseToSnakeCase
     </#list>
 </#if>
 
+
```
accidental newline ?

sruehl commented on build-utils/language-c/src/main/resources/templates/c/pojo-template-c.ftlh:
```diff
@@ -221,7 +227,7 @@ plc4c_return_code ${helper.getCTypeName(type.name)}_parse(plc4c_spi_read_buffer*
 <#if indentContent>  </#if>  {
 <#if indentContent>  </#if>    // Create an array of all the bytes read in this message element so far.
-<#if indentContent>  </#if>    byte[] checksumRawData = plc4c_spi_read_get_bytes(io, startPos, plc4c_spi_read_get_pos(io));
+<#if indentContent>  </#if>    //unsigned char checksumRawData[] = plc4c_spi_read_get_bytes(io, startPos, plc4c_spi_read_get_pos(io));
```
is that meant to be outcommeted?
```suggestion
<#if indentContent>  </#if>    unsigned char checksumRawData[] = plc4c_spi_read_get_bytes(io, startPos, plc4c_spi_read_get_pos(io));
```

hutcheb replied:
The checksum field stuff isn't supported in C yet, there is more to fix bit it was just something I saw.

sruehl commented on build-utils/language-c/src/test/resources/settings.xml:
```diff
@@ -0,0 +1,51 @@
+      </pluginRepositories>
+    </profile>
+  </profiles>
+</settings>
```
EOL at end of file
```suggestion
</settings>

```

sruehl commented on build-utils/language-go/src/main/java/org/apache/plc4x/language/go/GoLanguageTemplateHelper.java:
```diff
@@ -309,7 +309,7 @@ public String getReadBufferReadMethodCall(SimpleTypeReference simpleTypeReferenc
             case STRING: {
                 StringTypeReference stringTypeReference = (StringTypeReference) simpleTypeReference;
-                return "io.ReadString(" + stringTypeReference.getSizeInBits() + ")";
+                return "io.ReadString(" + stringTypeReference.getLength() + ")";
```
see other bit size vs byte site length comment

sruehl commented on build-utils/language-go/src/main/java/org/apache/plc4x/language/go/GoLanguageTemplateHelper.java:
```diff
@@ -368,7 +368,7 @@ public String getWriteBufferWriteMethodCall(SimpleTypeReference simpleTypeRefere
                 String encoding = ((stringTypeReference.getEncoding() != null) && (stringTypeReference.getEncoding().length() > 2)) ?
                     stringTypeReference.getEncoding().substring(1, stringTypeReference.getEncoding().length() - 1) : "UTF-8";
-                return "io.WriteString(" + stringTypeReference.getSizeInBits() + ", \"" +
+                return "io.WriteString(" + stringTypeReference.getLength() + ", \"" +
```
see other bit size vs byte site length comment

sruehl commented on build-utils/language-go/src/main/resources/templates/go/data-io-template.ftlh:
```diff
@@ -75,7 +75,8 @@ import (
 	<#list type.switchField.cases as case>
-		<#if case.discriminatorValues?has_content>case <#list case.discriminatorValues as discriminatorValue>${helper.toParseExpression(null, type.switchField.discriminatorExpressions[discriminatorValue?index], type.parserArguments)} == <#if helper.discriminatorValueNeedsStringEqualityCheck(type.switchField.discriminatorExpressions[discriminatorValue?index])>"${discriminatorValue}"<#elseif helper.isEnu ...
+        <#if case.discriminatorValues?has_content>case <#list case.discriminatorValues as discriminatorValue>${helper.toParseExpression(null, type.switchField.discriminatorExpressions[discriminatorValue?index], type.parserArguments)} == <#if helper.discriminatorValueNeedsStringEqualityCheck(type.switchField.discriminatorExpressions[discriminatorValue?index])>"${discriminatorValue}"<#elseif helper ...
+
```
accidental newline?
```suggestion
```

sruehl commented on build-utils/language-go/src/main/resources/templates/go/model-template.ftlh:
```diff
@@ -306,10 +322,18 @@ func (m *${type.name}) LengthInBits() uint16 {
 			<#case "simple">
 				<#assign simpleField = field>
 
+
```
Accidental newline(s)?
```suggestion
```

sruehl commented on build-utils/language-go/src/main/resources/templates/go/model-template.ftlh:
```diff
@@ -360,6 +384,7 @@ func ${type.name}Parse(io *utils.ReadBuffer<#if type.parserArguments?has_content
 					<#if (!helper.isSimpleTypeReference(arrayField.type)) && helper.requiresVariable(arrayField, "lastItem")>
 		lastItem := curItem == uint16(${helper.toParseExpression(arrayField, arrayField.loopExpression, type.parserArguments)} - 1)
 					</#if>
+
```
accidental newline?
```suggestion
```

sruehl commented on build-utils/language-go/src/main/resources/templates/go/model-template.ftlh:
```diff
@@ -306,10 +322,18 @@ func (m *${type.name}) LengthInBits() uint16 {
 				<#assign simpleField = field>
 
+
+
```
Accidental newline(s)?
```suggestion
```

sruehl commented on build-utils/language-go/src/main/resources/templates/go/model-template.ftlh:
```diff
@@ -306,10 +322,18 @@ func (m *${type.name}) LengthInBits() uint16 {
 
+
+
+
```
Accidental newline(s)?
```suggestion
```

sruehl commented on build-utils/language-go/src/test/resources/plc4go/pom.xml:
```diff
@@ -0,0 +1,323 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
was just wondering if we find a way to keep these in sync with the other poms or if we use tool native builds here.

sruehl replied:
mostly this is for me a discussion reminder. (pom is pretty big, also plc4go received some updates in the meantime, artifactid should be something like example.com:example e.g.)

hutcheb replied:
I have added a reference to the parent pom of build-utils for the integration tests. It would be nice to be able to keep it up to date with the plc4go pom, but not sure there is a way to do it.

At least for the integration tests I don't think there is much value a native tool as the Go build-utils is just a java project that happens to produce Go source code.

sruehl commented on build-utils/language-java/pom.xml:
```diff
@@ -55,4 +55,39 @@
+  </build>
+
+
 </project>
```
new line at EOF
```suggestion
</project>

```

sruehl commented on build-utils/language-java/src/main/java/org/apache/plc4x/language/java/JavaLanguageTemplateHelper.java:
```diff
@@ -363,7 +367,7 @@ public String getReadBufferReadMethodCall(SimpleTypeReference simpleTypeReferenc
             case STRING: {
                 StringTypeReference stringTypeReference = (StringTypeReference) simpleTypeReference;
-                return "io.readString(" + stringTypeReference.getSizeInBits() + ", \"" +
+                return "io.readString(" + stringTypeReference.getLength() + ", \"" +
```
see other bit length vs byte length comment

sruehl commented on build-utils/language-java/src/main/java/org/apache/plc4x/language/java/JavaLanguageTemplateHelper.java:
```diff
@@ -422,7 +425,7 @@ public String getWriteBufferWriteMethodCall(SimpleTypeReference simpleTypeRefere
             case STRING: {
                 StringTypeReference stringTypeReference = (StringTypeReference) simpleTypeReference;
-                return "io.writeString(" + stringTypeReference.getSizeInBits() + ", \"" +
+                return "io.writeString(" + stringTypeReference.getLength() + ", \"" +
```
see other bit length vs byte length comment

sruehl commented on build-utils/language-java/src/main/resources/templates/java/data-io-template.ftlh:
```diff
@@ -311,7 +313,9 @@ public class ${type.name}IO {
-        <#list type.switchField.cases as case><#if case.discriminatorValues?has_content>if(<#list case.discriminatorValues as discriminatorValue>EvaluationHelper.equals(${helper.toParseExpression(null, type.switchField.discriminatorExpressions[discriminatorValue?index], type.parserArguments)}, <#if helper.discriminatorValueNeedsStringEqualityCheck(type.switchField.discriminatorExpressions[discrim ...
+        <#list type.switchField.cases as case>
+            <#if case.discriminatorValues?has_content>if(<#list case.discriminatorValues as discriminatorValue>EvaluationHelper.equals(${helper.toParseExpression(null, type.switchField.discriminatorExpressions[discriminatorValue?index], type.parserArguments)},<#if helper.discriminatorValueNeedsStringEqualityCheck(type.switchField.discriminatorExpressions[discriminatorValue?index])>"${discriminato ...
+
```
```suggestion
```

accidental newline?

sruehl commented on build-utils/language-java/src/main/resources/templates/java/enum-template.ftlh:
```diff
@@ -59,7 +59,7 @@ import java.util.Map;
 
 <#list type.enumValues as enumValue>
-    ${enumValue.name}(<#if type.type?has_content>(${helper.getLanguageTypeNameForTypeReference(type.type, true)}) <#if helper.isStringTypeReference(type.type)>"${enumValue.value}"<#else>${enumValue.value}</#if></#if><#if type.constantNames?has_content><#if type.type?has_content>, </#if><#list type.constantNames as constantName>(${helper.getLanguageTypeNameForTypeReference(type.getConstantType(con ...
+    ${enumValue.name}(<#if type.type?has_content>(${helper.getLanguageTypeNameForTypeReference(type.type, true)}) <#if helper.isStringTypeReference(type.type)>"${enumValue.value}"<#elseif helper.isComplexTypeReference(type.type)><#if helper.isEnumTypeReference(type.type)>${helper.getLanguageTypeNameForTypeReference(type.type, true)}.${enumValue.value}<#else>${enumValue.value}</#if><#else>${enumVa ...
```
general comment: we should research if we can add some line breaks here 😄

sruehl replied:
in all other places where we get such long lines too 😉

hutcheb replied:
Yeah these long line suck.

I'm guessing we would probably have to implement a helper function or there looks like the might be a <#compress> directive that might be helpful. Let me check.

hutcheb replied:
`  {/* ${helper.getCTypeName(name)} */
<#if true>   </#if><@compress single_line=true>
               <#list typeValues as key, value>
                   <#if !helper.isComplexTypeReference(helper.getDiscriminatorTypes()[key])>
                .${key} = <#if value??>${value}<#else>-1</#if><#sep>,
                   <#else>
                .${key} = <#if value??>${helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[key])}_${value}<#else>-1</#if><#sep>,
                   </#if></#list>}<#sep >,
            </@compress>`

Something like that seems to work ok anything between the compress tags strips all multiple white space/line breaks to a single space, @chrisdutz what do you think?

chrisdutz replied:
Sounds good.

sruehl commented on build-utils/language-java/src/main/resources/templates/java/io-template.ftlh:
```diff
@@ -246,7 +246,12 @@ public class ${type.name}IO implements <#if outputFlavor != "passive">MessageIO<
+        <#else>
+        ${helper.getLanguageTypeNameForField(field)} ${discriminatorField.name} = <#if helper.isSimpleTypeReference(discriminatorField.type)>${helper.getReadBufferReadMethodCall(discriminatorField.type)}<#else>${discriminatorField.type.name}IO.staticParse(io<#if field.params?has_content>, <#list field.params as parserArgument>(${helper.getLanguageTypeNameForTypeReference(helper.getArgumentType(di ...
+        </#if>
+
```
```suggestion
```
accidental newline?

sruehl commented on build-utils/language-java/src/test/resources/integration-test/pom.xml:
```diff
@@ -0,0 +1,224 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
same as other comment about test-pom.xml: is there a way to keep them easy in sync

sruehl commented on build-utils/protocol-base-mspec/src/main/java/org/apache/plc4x/plugins/codegenerator/language/mspec/model/definitions/DefaultDataIoTypeDefinition.java:
```diff
@@ -21,18 +21,24 @@ Licensed to the Apache Software Foundation (ASF) under one
         return switchField;
     }
 
+    public TypeReference getType() { return this.type; }
```
```suggestion
    public TypeReference getType() { 
        return this.type; 
    }
```

sruehl commented on build-utils/protocol-base-mspec/src/main/java/org/apache/plc4x/plugins/codegenerator/language/mspec/parser/MessageFormatListener.java:
```diff
@@ -42,6 +42,8 @@ Licensed to the Apache Software Foundation (ASF) under one
 
 public class MessageFormatListener extends MSpecBaseListener {
 
+
```
```suggestion
```
accidental newlines?

sruehl commented on build-utils/protocol-base-mspec/src/main/java/org/apache/plc4x/plugins/codegenerator/language/mspec/parser/MessageFormatListener.java:
```diff
@@ -42,6 +42,8 @@ Licensed to the Apache Software Foundation (ASF) under one
 public class MessageFormatListener extends MSpecBaseListener {
 
+
+
```
```suggestion
```
accidental newlines?

sruehl commented on build-utils/protocol-base-mspec/src/test/java/org/apache/plc4x/plugins/codegenerator/language/mspec/parser/MessageFormatParserTest.java:
```diff
@@ -39,5 +41,6 @@ void parseNull() {
     void parseSomething() {
         Map<String, TypeDefinition> parse = SUT.parse(getClass().getResourceAsStream("/mspec.example"));
         System.out.println(parse);
+
```
accidental newline?
```suggestion
```

sruehl commented on build-utils/protocol-base-mspec/src/test/java/org/apache/plc4x/plugins/codegenerator/language/mspec/parser/MessageFormatParserTest.java:
```diff
@@ -39,5 +41,6 @@ void parseNull() {
         System.out.println(parse);
+
     }
 }
```
newline at EOF
```suggestion
}

```

sruehl commented on build-utils/protocol-base-mspec/src/test/resources/mspec.example:
```diff
@@ -208,9 +208,9 @@
-        ['CommandId.INVALID', 'true' AdsInvalidResponse]
-        ['CommandId.INVALID', 'false' AdsInvalidRequest]
-        ['CommandId.ADS_READ_DEVICE_INFO', 'true' AdsReadDeviceInfoResponse
+        ['INVALID', 'true' AdsInvalidResponse]
```
Think @chrisdutz talked about that: This is still a reference to a existing enum not a string right? Like you don't need to qualify them here

sruehl commented on build-utils/protocol-base-mspec/src/test/resources/mspec.example:
```diff
@@ -355,4 +355,11 @@
+    //Confirm implicit can be used in string length
+    [implicit uint 32 'stringLength' 'stringValue.lengthInBytes']
+    [simple string 'stringLength' 'stringValue']
 ]
```
newline at EOF
```suggestion
]

```

sruehl commented on build-utils/protocol-test/pom.xml:
```diff
@@ -74,13 +215,6 @@
-    </dependency>
   </dependencies>
 
 </project>
```
newline at EOF
```suggestion
</project>

```

sruehl commented on build-utils/protocol-test/src/main/resources/protocols/test/test.mspec:
```diff
@@ -21,50 +21,319 @@
 // Simple Type
```
Maybe add a comment about the purpose of this file here like:
```suggestion
////////////////////////////////////////////////////////////////
// This file is a crossproduct of all mspec features
////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////
// Simple Type
```

sruehl commented on build-utils/protocol-test/src/main/resources/protocols/test/test.mspec:
```diff
@@ -21,50 +21,319 @@
+            [simple uint 16 'value']
+        ]
+    ]
+]
```
newline at EOF
```suggestion
]

```

sruehl commented on go.mod:
```diff
@@ -19,3 +19,5 @@
 module github.com/apache/plc4x
 
 go 1.15
+
```
This toplevel go.mod file confuses me. What is the purpose of that? I open plc4go as TLP in golang so maybe this is from opening the whole plc4x in goland?

sruehl commented on go.sum:
```diff
@@ -0,0 +1,16 @@
+github.com/ajankovic/xdiff v0.0.1/go.mod h1:SUmEZ67uB97I0zkiuQ+lb+LOms9ipn8X+p+2RdJV710=
```
see go.mod comment above

splatch commented on build-utils/language-c/src/main/java/org/apache/plc4x/language/c/CLanguageTemplateHelper.java:
```diff
@@ -178,33 +178,44 @@ public boolean requiresPointerAccess(ComplexTypeDefinition typeDefinition, Strin
      */
     @Override
     public String getLanguageTypeNameForTypeReference(TypeReference typeReference) {
+        System.out.println("Type Reference " + typeReference.getClass());
```
Small leftover to be cleaned up.

sruehl replied:
(the fix is contained in my mega review below 😉 )

sruehl commented on plc4go/internal/plc4go/ads/readwrite/model/AdsData.go:
```diff
@@ -98,45 +98,65 @@ func AdsDataParse(io *utils.ReadBuffer, commandId *CommandId, response bool) (*A
 	var typeSwitchError error
 	switch {
-	case *commandId == CommandId_INVALID && response == false:
+
```
Guess all these are artifacts from the superfluous newlines...

splatch commented on build-utils/language-java/src/main/java/org/apache/plc4x/language/java/JavaLanguageTemplateHelper.java:
```diff
@@ -26,13 +26,17 @@ Licensed to the Apache Software Foundation (ASF) under one
 @SuppressWarnings({"unused", "WeakerAccess"})
 public class JavaLanguageTemplateHelper extends BaseFreemarkerLanguageTemplateHelper {
 
+    private final Logger logger = LoggerFactory.getLogger(JavaLanguageTemplateHelper.class);
```
I see new logger field but can't see direct use of it.

sruehl replied:
also it should be static and upper case for sonar 😄

splatch reviewed (approved):
I love automated tests of mspec. That's huge contribution. My comments are only for java about minor things. They should not hold this PR from coming in.

sruehl commented on pom.xml:
```diff
@@ -100,7 +100,7 @@
     <sonar.exclusions>**/generated-sources</sonar.exclusions>
 
-    <plc4x-code-generation.version>1.4.0</plc4x-code-generation.version>
+    <plc4x-code-generation.version>1.5.0-SNAPSHOT</plc4x-code-generation.version>
```
uh are there some new upstream feature required?

hutcheb replied:
Yes there is an feature/native_opcua branch for the build-utils that includes some changes to the string type references.

sruehl replied:
@chrisdutz do we need to release the code-generation stuff first or can we leave it at SNAPSHOT

sruehl commented on protocols/ads/src/main/resources/protocols/ads/ads.mspec:
```diff
@@ -386,7 +386,7 @@
 ]
 
-[dataIo 'DataItem' [string 'dataFormatName', int 32 'stringLength']
+[dataIo 'DataItem' [string '-1' 'dataFormatName', int 32 'stringLength']
```
what does '-1' mean for a string?

hutcheb replied:
'-1' means that it is a null terminated string, which was the default previously.
We can't have an optional field here using expression as we already have the encoding field straight after it.

sruehl reviewed (commented):
LGTM: just some minor formatting stuff and some general questions 😄

hutcheb committed:
Commmit before refactoring

hutcheb committed:
Refactored the String length expression to a Term.

Cleaned up a little bit.

C and .NET still left to do

hutcheb committed:
Updated the C branch with the String length expression.

updated pom.xml to inherit stuff for the integration tests
Updated the RAT and gitignore to only ignore C Makefiles as these
are automatically generated from cmake.

chrisdutz commented on build-utils/language-c/src/test/resources/integration-test/pom.xml:
```diff
@@ -324,65 +312,63 @@
     <groupId>org.apache.plc4x</groupId>
     <artifactId>plc4j-api</artifactId>
-    <version>0.9.0-SNAPSHOT</version>
+    <version>${project.version}</version>
```
Why did you change this? The maven release plugin takes care of updating this.

chrisdutz commented on build-utils/language-go/src/test/resources/plc4go/pom.xml:
```diff
@@ -244,75 +225,61 @@
     <groupId>org.apache.plc4x</groupId>
     <artifactId>plc4j-api</artifactId>
-    <version>0.9.0-SNAPSHOT</version>
+    <version>${project.version}</version>
```
Please leave this at the normal version not a variable.

hutcheb replied:
Changed back.

hutcheb committed:
Fixed various formatting issues.

hutcheb committed:
Merge branch 'develop' into feature/string_enum_mspec

hutcheb committed:
Removed the local repository requirements for the build-util modules.

hutcheb committed:
Minor change

sruehl commented on .gitignore:
```diff
@@ -181,3 +181,14 @@ build-iPhoneSimulator/
+*.includecache
+CMakeFiles/
+CMakeCache.txt
+DartConfiguration.tcl
```
```suggestion
DartConfiguration.tcl

```

sruehl commented on build-utils/language-base-freemarker/pom.xml:
```diff
@@ -55,6 +55,12 @@
+      </dependency>
   </dependencies>
 
 </project>
```
```suggestion
</project>

```

sruehl commented on build-utils/language-java/src/main/resources/templates/java/pojo-template.ftlh:
```diff
@@ -95,7 +95,11 @@ public<#if helper.isDiscriminatedParentTypeDefinition()> abstract</#if> class ${
     public ${helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])} get${discriminatorName?cap_first}() {
-        return <#if discriminatorValue??>${helper.adjustLiterals(helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName]),discriminatorValue)}<#else>${helper.getNullValueForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}</#if>;
+            <#if !helper.isComplexTypeReference(helper.getDiscriminatorTypes()[discriminatorName])>
+                return <#if helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName]) = "String"><#if discriminatorValue??>"${discriminatorValue}"<#else>${helper.getNullValueForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}</#if><#else><#if discriminatorValue??>${helper.adjustLiterals(helper.getLanguageTypeNameForTypeReference(helper.getDiscr ...
```
```suggestion
        return <#if helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName]) = "String"><#if discriminatorValue??>"${discriminatorValue}"<#else>${helper.getNullValueForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}</#if><#else><#if discriminatorValue??>${helper.adjustLiterals(helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName]),discriminatorValue)}<#else>${helper.getNullValueForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}</#if></#if>;
```

sruehl commented on build-utils/language-java/src/main/resources/templates/java/pojo-template.ftlh:
```diff
@@ -95,7 +95,11 @@ public<#if helper.isDiscriminatedParentTypeDefinition()> abstract</#if> class ${
+            <#if !helper.isComplexTypeReference(helper.getDiscriminatorTypes()[discriminatorName])>
+                return <#if helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName]) = "String"><#if discriminatorValue??>"${discriminatorValue}"<#else>${helper.getNullValueForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}</#if><#else><#if discriminatorValue??>${helper.adjustLiterals(helper.getLanguageTypeNameForTypeReference(helper.getDiscr ...
+            <#else>
+                return <#if helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName]) = "String"><#if discriminatorValue??>"${discriminatorValue}"<#else>${helper.getNullValueForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}</#if><#else><#if discriminatorValue??>${helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discri ...
```
```suggestion
        return <#if helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName]) = "String"><#if discriminatorValue??>"${discriminatorValue}"<#else>${helper.getNullValueForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}</#if><#else><#if discriminatorValue??>${helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}.${discriminatorValue}<#else>${helper.getLanguageTypeNameForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}.${helper.getNullValueForTypeReference(helper.getDiscriminatorTypes()[discriminatorName])}</#if></#if>;
```
